### DISCUSSION TITLE
Add a way of getting the Item entities dropped when a player dies

### DISCUSSION THREAD

**Discussion by WiIIiam278** [PaperMC/Paper#10603] [Ideas]
### Is your feature request related to a problem?

I'd like to be able to access and edit the _item entities_ a player drops when they die (not just the ItemStacks) - for example, to [mark them as "owned"](https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/entity/Item.html#setOwner(java.util.UUID)) by another user, changing their despawn time, moving them in the game world, etc - things not possible with the itemstack list exposed by the death event.

This was possible for a time with https://github.com/PaperMC/Paper/pull/7650. However, this change was reverted due to plugin incompatibilities (https://github.com/PaperMC/Paper/pull/10046).

### Describe the solution you'd like.

Consider introducing a (Player/Entity)DropItemOnDeath, or similar, to avoid breaking plugin compatibility while providing API to access item drops entities when they spawn in.

EDIT: A suggestion by molor was to add a method returning a nullable/optional `Entity` into `ItemSpawnEvent`, indicating the entity that caused the item to spawn, and a method indicating _how_ the item spawned (a `Cause` enum?)

### Describe alternatives you've considered.

The workaround for doing stuff with item death entities at the minute involves:

* Listening to the PlayerDeathEvent with a high priority (to handle other plugins modifying the drops)
* Marking or saving the dropped ItemStacks in a list
* Iterate through all item entities within a radius of where the player died and checking against the saved list / marked item meta (in a delay task to account for the ticks it takes the items to spawn)

Which is quite clunky and expensive, and not guaranteed to be instantaneous when the items spawn.

**Comment by molor** [comment]
Why only for `Player`? I think it would be more useful to have more a generic event like `EntityDeathDropItemEvent`, at least because I also needed it yesterday (and because in `(Item/Entity)SpawnEvent` it is currently impossible to get which mob's death led to the spawn of this `Item` entity).

Another idea is to add method like `getCauseEntity()` to `ItemSpawnEvent` which will return the `Entity` that led to the spawn of this `Item` (a dead entity, or the entity who threw it (player (from inventory) or mob (i.e. zombie when picking better armor)...)

**Comment by WiIIiam278** [comment]
> Why only for `Player`? (...)

Yeah, fair. I'll amend the issue description.

> add method like `getCauseEntity()` to `ItemSpawnEvent` (...)

~~Issue is that any changes to how often the ItemSpawnEvent is fired breaks compatibility quite significantly, apparently:
https://github.com/PaperMC/Paper/issues/10044~~

EDIT: I misread. This actually seems like a sensible idea, maybe? You'd still need a way of knowing _how_ the entity spawned the item (I only want the items spawned by a player dropping them). Perhaps a Cause enum?

**Comment by WiIIiam278** [comment]
Just noticed this was implemented in https://github.com/PaperMC/Paper/pull/7332, but was closed due to being stale :(

**Comment by NonSwag** [comment]
I found a workaround for that
When the player dies I add all items to a map and on item spawn get them back with the player as the value
This is not optimal but a temporary solution
